src/gui/plugins/mouse_drag/MouseDrag.cc

```diff
@@ -0,0 +1,830 @@
+      planeMat->SetDepthCheckEnabled(false);
+      planeMat->SetDepthWriteEnabled(false);
+    }
+
+    this->planeVisual = scene->CreateVisual();
+    this->planeVisual->AddGeometry(scene->CreatePlane());
+    this->planeVisual->SetInheritScale(false);
+    this->planeVisual->SetMaterial(planeMat);
+    this->planeVisual->SetUserData("gui-only", static_cast<bool>(true));
+  }
+
+  // Update the visualization
+  if (!this->dragActive)
+  {
+    this->arrowVisual->SetVisible(false);
+    this->boxVisual->SetVisible(false);
+    this->planeVisual->SetVisible(false);
+  }
+  else if (this->mode == MouseDragMode::ROTATE)
+  {
+    this->boxVisual->SetLocalPosition(this->applicationPoint);
+    this->boxVisual->SetLocalRotation(this->goalRot);
+    this->boxVisual->SetLocalScale(1.2 * this->bboxSize);
+
+    this->arrowVisual->SetVisible(false);
+    this->boxVisual->SetVisible(true);
+    this->planeVisual->SetVisible(false);
+  }
+  else if (this->mode == MouseDragMode::TRANSLATE)
+  {
+    const math::Vector3d axisDir = this->target - this->applicationPoint;
+    math::Quaterniond quat;
+    quat.SetFrom2Axes(math::Vector3d::UnitZ, axisDir);
+    double scale = 2 * this->bboxSize.Length();
+    this->arrowVisual->SetLocalPosition(this->applicationPoint);
+    this->arrowVisual->SetLocalRotation(quat);
+    this->arrowVisual->SetLocalScale(scale, scale, axisDir.Length() / 0.75);
+
+    quat.SetFrom2Axes(math::Vector3d::UnitZ, -this->plane.Normal());
+    scale = this->applicationPoint.Distance(this->camera->WorldPose().Pos());
+    this->planeVisual->SetLocalRotation(quat);
+    this->planeVisual->SetLocalPosition(this->applicationPoint);
+    this->planeVisual->SetLocalScale(scale);
+
+    this->arrowVisual->SetVisible(true);
+    this->boxVisual->SetVisible(false);
+    this->planeVisual->SetVisible(true);
+  }
+
+  if (this->sendBlockOrbit)
+  {
+    // Events with false should only be sent once
+    if (!this->blockOrbit)
+      this->sendBlockOrbit = false;
+
+    gz::gui::events::BlockOrbit blockOrbitEvent(this->blockOrbit);
+    gz::gui::App()->sendEvent(
+        gz::gui::App()->findChild<gz::gui::MainWindow *>(),
+        &blockOrbitEvent);
+  }
+}
+
+/////////////////////////////////////////////////
+void MouseDragPrivate::HandleMouseEvents()
+{
+  // Check for mouse events
+  if (!this->mouseDirty)
+  {
+    return;
+  }
+  this->mouseDirty = false;
+
+  std::lock_guard<std::mutex> lock(this->mutex);
+
+  if (this->mouseEvent.Type() == common::MouseEvent::PRESS &&
+      this->mouseEvent.Control() &&
+      this->mouseEvent.Button() != common::MouseEvent::MIDDLE)
+  {
+    // Get the visual at mouse position
+    const rendering::VisualPtr visual = this->scene->VisualAt(
+      this->camera,
+      this->mouseEvent.Pos());
+
+    if (!visual)
+    {
+      this->mode = MouseDragMode::NONE;
+      return;
+    }
+    try
+    {
+      this->linkId =
+        std::get<uint64_t>(visual->Parent()->UserData("gazebo-entity"));
```

Comment:
I got a segfault while testing this and I think it's because `visual->Parent()` was `nullptr`. Not sure how I got it to do that, but we should check first.

Here's the backtrace:
```
#0  __pthread_kill_implementation (no_tid=0, signo=11, threadid=139740024129088) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=11, threadid=139740024129088) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=139740024129088, signo=signo@entry=11) at ./nptl/pthread_kill.c:89
#3  0x00007f1a79642476 in __GI_raise (sig=11) at ../sysdeps/posix/raise.c:26
#4  0x00007f1a79d86a30 in backward::SignalHandling::sig_handler(int, siginfo_t*, void*) (signo=<optimized out>, info=0x7f17c27faef0, _ctx=<optimized out>) at ~/ws/garden/src/gz-tools/src/Backward/backward.hpp:4268
#5  0x00007f1a79642520 in <signal handler called> () at /lib/x86_64-linux-gnu/libc.so.6
#6  gz::sim::MouseDragPrivate::HandleMouseEvents() (this=0x55d9814481c0) at ~/ws/garden/src/gz-sim/src/gui/plugins/mouse_drag/MouseDrag.cc:677
#7  0x00007f1a2c2cd267 in gz::sim::MouseDrag::eventFilter(QObject*, QEvent*) (this=0x55d980f019f0, _obj=0x55d97d425ab0, _event=0x7f17c27fb5a0) at ~/ws/garden/src/gz-sim/src/gui/plugins/mouse_drag/MouseDrag.cc:314
#8  0x00007f1a700b9b9a in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007f1a70d6c702 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#10 0x00007f1a700b9e3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007f1a40568f34 in gz::gui::plugins::GzRenderer::BroadcastMousePress() (this=0x55d97dff8bb0) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:523
#12 0x00007f1a40568bbd in gz::gui::plugins::GzRenderer::HandleMouseEvent() (this=0x55d97dff8bb0) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:388
#13 0x00007f1a40568695 in gz::gui::plugins::GzRenderer::Render(gz::gui::plugins::RenderSync*) (this=0x55d97dff8bb0, _renderSync=0x55d97dff8a50) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:336
#14 0x00007f1a40590015 in gz::gui::plugins::RenderThreadRhiOpenGL::RenderNext(gz::gui::plugins::RenderSync*) (this=0x55d97dff8f50, _renderSync=0x55d97dff8a50) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalSceneRhiOpenGL.cc:158
#15 0x00007f1a4056ba91 in gz::gui::plugins::RenderThread::RenderNext(gz::gui::plugins::RenderSync*) (this=0x55d97dff8b80, _renderSync=0x55d97dff8a50) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:745
#16 0x00007f1a4058dc56 in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<gz::gui::plugins::RenderSync*>, void, void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderSync*)>::call(void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderSync*), gz::gui::plugins::RenderThread*, void**) (f=(void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderThread * const, gz::gui::plugins::RenderSync *)) 0x7f1a4056ba60 <gz::gui::plugins::RenderThread::RenderNext(gz::gui::plugins::RenderSync*) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:744>, o=0x55d97dff8b80, arg=0x7f18a6521918) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:152
#17 0x00007f1a4058dbbd in QtPrivate::FunctionPointer<void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderSync*)>::call<QtPrivate::List<gz::gui::plugins::RenderSync*>, void>(void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderSync*), gz::gui::plugins::RenderThread*, void**) (f=(void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderThread * const, gz::gui::plugins::RenderSync *)) 0x7f1a4056ba60 <gz::gui::plugins::RenderThread::RenderNext(gz::gui::plugins::RenderSync*) at ~/ws/garden/src/gz-gui/src/plugins/minimal_scene/MinimalScene.cc:744>, o=0x55d97dff8b80, arg=0x7f18a6521918) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:185
#18 0x00007f1a4058db02 in QtPrivate::QSlotObject<void (gz::gui::plugins::RenderThread::*)(gz::gui::plugins::RenderSync*), QtPrivate::List<gz::gui::plugins::RenderSync*>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x7f18a4389090, r=0x55d97dff8b80, a=0x7f18a6521918, ret=0x0) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:418
#19 0x00007f1a700e741e in QObject::event(QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#20 0x00007f1a70d6c713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#21 0x00007f1a700b9e3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#22 0x00007f1a700bcf27 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#23 0x00007f1a70113a67 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#24 0x00007f1a6e935d3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#25 0x00007f1a6e98b258 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007f1a6e9333e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#27 0x00007f1a701130b8 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#28 0x00007f1a700b875b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#29 0x00007f1a6fecbaf2 in QThread::exec() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#30 0x00007f1a6feccca1 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#31 0x00007f1a79694b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#32 0x00007f1a79726a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

Reply:
Done in 3627480